R/checkUnitFactor.R

```diff
@@ -43,7 +43,16 @@ checkUnitFactor <- function(template, logFile = NULL, failOnUnitMismatch = TRUE)
                           c("1.17", "EUR_2020", "US$2005"),
                           c("1.174", "EUR2020", "US$2005"),
                           # temporary, error in ARIADNE mapping for 'Capital Stock'
-                          c("1.174", "billion EUR2020/yr", "billion US$2005")
+                          c("1.174", "billion EUR2020/yr", "billion US$2005"),
+                          # AgMIP
+                          c("1000", "1000 ha", "million ha"),
+                          c("1", "MtCO2", "Mt CO2/yr"),
+                          c("1000", "MtCO2e", "Gt CO2e/yr"),
+                          c("1", "MtCO2e", "Mt CO2e/yr"),
+                          c("1000", "1000 t", "Mt Nr/yr"),
+                          c("1", "TgN/year", "Mt Nr/yr"),
+                          c("0.001", "bn USD 2005 MER", "million US$05 MER/yr"),
+                          c("1000", "1000 t dm", "Mt DM/yr")

```

Comment:
I'm not happy with all of these additions, to be honest.

A `Mt CO2/yr` is not the same as `MtCO2`. I think these are errors in the template of AgMIP and should better be corrected there. The values are clearly in `Mt/yr` and using `Mt` here is both scientifically wrong and potentially confusing, in particular as these reportings often also contain variables in `Mt CO2` such as cumulative emissions. Same with the `t/yr` and similar ones. Of course, I have absolutely no say in the AgMIP project, but you might want to raise it with the persons responsible

We added these checks because we had quite a number of cases where things with inconsistent units were mapped. If you now "accept" them by adding them here, we will not be able to catch future wrong mappings regarding this unit anymore. If every project just adds the (in my opinion wrong) relations here, this will become quite useless in the future.

If the `AgMIP` template cannot be adjusted anymore, this is of course the way to go.

Reply:
Agree that it would be bad for this to become a dumping ground for such inconsistencies in every possible mapping as this is a shared list between different templates and don't want to set a precedent that erodes this check.   

We now plan on having the mapping keep that years here and handle the issue either in the output script that calls piamInterfaces or via the submission tool for AgMIP. That will work but I guess splits the work on mapping across multiple places and would be nice to be able to handle it all in one place. The units may be changeable after discussion with the project but until then will have to do the correction somewhere.

Another thing I can think of is what about adding an optional parameter into the checkUnitFactor function that allows to pass template specific exceptions? Then these won't be an issue for other projects and known inconsistencies can be tracked?

Reply:
I like the idea of template-specific exceptions, as they make sources explicit and reduce the possibility of unwanted side effects.